### PR TITLE
WebSocket Observer connector

### PR DESCRIPTION
**Pull request by ssd04**
Web socket connector for observer node communication.

TODO:
- Old flow will be removed/refactored in a following PR
- Integration tests will be updated in a following PR

### CODE REVIEW

File: cmd/notifier/config/config.toml [hunk 1]
```diff
@@ -13,6 +13,26 @@
         Prefix = "erd"
         Length = 32
 
+[WebSocketConnector]
+    # URL for the WebSocket client/server connection
+    # This value represents the IP address and port number that the WebSocket client or server will use to establish a connection.
+    url = "localhost:22111"
+
+    # This flag describes the mode to start the WebSocket connector. Can be "client" or "server"
+    mode = "server"
```
Comment (bogdan-rosianu) [inline cmd/notifier/config/config.toml:22]: why not client?

Comment (bogdan-rosianu) [reply cmd/notifier/config/config.toml:22]: also, keep consistency for the TOML properties. If the old ones were following the uppercase + camel case pattern, do the same for the new ones (so, for example, `data-marshalller-type` -> `DataMarshallerType`)

File: process/dataPreProcessor.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	coreData "github.com/multiversx/mx-chain-core-go/data"
+	"github.com/multiversx/mx-chain-core-go/data/block"
+	"github.com/multiversx/mx-chain-core-go/data/outport"
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	"github.com/multiversx/mx-chain-notifier-go/common"
+	"github.com/multiversx/mx-chain-notifier-go/data"
+)
+
+// ErrNilBlockData signals that a nil block data has been provided
+var ErrNilBlockData = errors.New("nil block data")
```
Comment (bogdan-rosianu) [inline process/dataPreProcessor.go:18]: could have defined a `var ( .. )` group

File: process/dataPreProcessor.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+// RevertIndexedBlock will handle the revert block event
+func (d *eventsDataPreProcessor) RevertIndexedBlock(blockData *outport.BlockData) error {
+	header, err := d.getHeaderFromBytes(core.HeaderType(blockData.HeaderType), blockData.HeaderBytes)
+	if err != nil {
+		return err
+	}
+
+	revertData := &data.RevertBlock{
+		Hash:  string(header.GetRootHash()),
```
Comment (bogdan-rosianu) [inline process/dataPreProcessor.go:141]: do you use string of bytes everywhere? We had issues in the past when using string(bytes) vs hex encode(bytes)

File: cmd/notifier/config/config.toml (hunk 1 above)
Comment (ssd04) [reply cmd/notifier/config/config.toml]: we use notifier as server, and observer nodes as clients; the same pattern was before with http requests

File: process/dataPreProcessor.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+	err = container.Add(core.MetaHeader, block.NewEmptyMetaBlockCreator())
+	if err != nil {
+		return nil, err
+	}
+
+	return container, nil
+}
+
+func (d *eventsDataPreProcessor) IsInterfaceNil() bool {
```
Comment (miiu96) [inline process/dataPreProcessor.go:187]: missing comment

File: process/eventsIndexer.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		outport.TopicSaveValidatorsRating:  ei.saveValidatorsRating,
+		outport.TopicSaveValidatorsPubKeys: ei.saveValidatorsPubKeys,
+		outport.TopicSaveAccounts:          ei.saveAccounts,
+		outport.TopicFinalizedBlock:        ei.finalizedBlock,
+	}
+}
+
+// ProcessPayload will proces the provided payload based on the topic
+func (ei *eventsIndexer) ProcessPayload(payload []byte, topic string) error {
+	payloadTypeAction, ok := ei.actions[topic]
```
[the remark this answers is not in the document]
Comment (ssd04) [reply process/eventsIndexer.go:57]: changed to not return error

File: process/dataPreProcessor.go (hunk 4 above)
Comment (ssd04) [reply process/dataPreProcessor.go:187]: added comment

**Review by bogdan-rosianu — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Pull request merged**